framework/audio/internal/audiobuffer.cpp

```diff
@@ -0,0 +1,110 @@
+//  it under the terms of the GNU General Public License version 2.
+//
+//  This program is distributed in the hope that it will be useful,
+//  but WITHOUT ANY WARRANTY; without even the implied warranty of
+//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+//  GNU General Public License for more details.
+//
+//  You should have received a copy of the GNU General Public License
+//  along with this program; if not, write to the Free Software
+//  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
+//=============================================================================
+#include "audiobuffer.h"
+#include <cstring>
+#include "log.h"
+
+using namespace mu::audio;
+
+AudioBuffer::AudioBuffer(unsigned int streamsPerSample, unsigned int size)
+    : m_streamsPerSample(streamsPerSample), m_writepointer(0), m_readpointer(0), m_data(), m_source(nullptr)
+{
+    m_data.resize(size * m_streamsPerSample, 0.f);
+}
+
+void AudioBuffer::setSource(std::shared_ptr<IAudioSource> source)
+{
+    m_source = source;
+}
+
+void AudioBuffer::fillup()
+{
+    IF_ASSERT_FAILED(m_source) {
+        return;
+    }
+
+    while (sampleLag() < MAX_SAMPLELAG) {
+        m_source->setBufferSize(FILL_SAMPLES);
+        m_source->forward(FILL_SAMPLES);
+        push(m_source->data(), FILL_SAMPLES);
+    }
+}
+
+void AudioBuffer::push(const float* source, int sampleCount)
+{
+    auto from = m_writepointer;
+    auto memStep = sizeof(float);
+    auto to = m_writepointer + sampleCount * m_streamsPerSample;
+    if (to > m_data.size()) {
+        to = m_data.size() - 1;
+    }
+    auto count = to - from;
+    std::memcpy(m_data.data() + m_writepointer, source, count * memStep);
+    m_writepointer += count;
+
+    int left = sampleCount * m_streamsPerSample - count;
+    if (left > 0) {
+        std::memcpy(m_data.data(), source + count, left * memStep);
+        m_writepointer = left;
+    }
+
+    if (m_writepointer >= m_data.size()) {
+        m_writepointer -= m_data.size();
+    }
+}
+
+void AudioBuffer::pop(float* dest, int sampleCount)
+{
+    if (sampleLag() < MIN_SAMPLELAG) {
+        std::memset(dest, 0, sampleCount * m_streamsPerSample);
+        return;
+    }
+
+    auto from = m_readpointer;
+    auto memStep = sizeof(float);
+    auto to = m_readpointer + sampleCount * m_streamsPerSample;
+    if (to > m_data.size()) {
+        to = m_data.size();
+    }
+    auto count = to - from;
+    std::memcpy(dest, m_data.data() + from, count * memStep);
+    m_readpointer += count;
+
+    int left = sampleCount * m_streamsPerSample - count;
+    if (left > 0) {
+        std::memcpy(dest + count, m_data.data(), left * memStep);
+        m_readpointer = left;
+    }
+
+    if (m_readpointer >= m_data.size()) {
+        m_readpointer -= m_data.size();
+    }
+}
+
+unsigned int AudioBuffer::sampleLag() const
+{
+    unsigned int lag = 0;
+    if (m_readpointer <= m_writepointer) {
```

Comment:
not thread-safe (accessed from different threads)

Reply:
@igorkorsukov indexes are atomic.